src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +137,132 @@ private void logTargetPath() {
+    }
+
+    /**
+     * Updates the estimated pose from the cameras.
+     */
     private void updateFromVision() {
-        getViableVisionObservations().stream()
-                .sorted(Comparator.comparingDouble(PoseEstimator6328.VisionObservation::timestamp))
-                .forEach(poseEstimator6328::addVisionObservation);
+        Arrays.stream(aprilTagCameras).forEach(AprilTagCamera::update);
+
+        final AprilTagCamera[] newResultCameras = Arrays.stream(aprilTagCameras)
+                .filter(AprilTagCamera::hasNewResult)
+                .toArray(AprilTagCamera[]::new);
+        sortCamerasByLastTargetTimestamp(newResultCameras);
+
+        updateEstimatedPoseFromVision();
     }
 
-    private List<PoseEstimator6328.VisionObservation> getViableVisionObservations() {
-        List<PoseEstimator6328.VisionObservation> viableVisionObservations = new ArrayList<>();
+    private void sortCamerasByLastTargetTimestamp(AprilTagCamera[] aprilTagCameras) {
+        QuickSort.sort(aprilTagCameras, AprilTagCamera::getLatestResultTimestampSeconds);
+    }
+
+    /**
+     * Sets the estimated pose from each camera, each at the timestamp of their latest result.
+     */
+    private void updateEstimatedPoseFromVision() {
         for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
-            final PoseEstimator6328.VisionObservation visionObservation = getVisionObservation(aprilTagCamera);
-            if (visionObservation != null)
-                viableVisionObservations.add(visionObservation);
+            if (!aprilTagCamera.hasNewResult())
+                continue;
+
+            addVisionObservation(
+                    aprilTagCamera.getEstimatedRobotPose(),
+                    aprilTagCamera.calculateStandardDeviations(),
+                    aprilTagCamera.getLatestResultTimestampSeconds()
+            );
         }
-        return viableVisionObservations;
     }
 
-    private PoseEstimator6328.VisionObservation getVisionObservation(AprilTagCamera aprilTagCamera) {
-        aprilTagCamera.update();
-        if (!aprilTagCamera.hasNewResult())
-            return null;
-        final Pose2d robotPose = aprilTagCamera.getEstimatedRobotPose();
-        if (robotPose == null || robotPose.getTranslation() == null || robotPose.getRotation() == null)
-            return null;
+    private void addVisionObservation(Pose2d estimatedPose, PoseEstimatorConstants.StandardDeviations standardDeviations, double timestamp) {
+        if (isVisionObservationTooOld(timestamp))
+            return;
+
+        final Pose2d poseAtObservationTimestamp = getPoseAtTimestamp(timestamp);
+        if (poseAtObservationTimestamp == null)
+            return;
+
+        final Transform2d odometryPoseToObservationTimestampPoseTransform = new Transform2d(odometryPose, poseAtObservationTimestamp);
+        final Pose2d estimatedPoseAtObservationTime = estimatedPose.plus(odometryPoseToObservationTimestampPoseTransform);
```

Comment:
this is just what getPoseAtTimestamp does, no?

Reply:
The getPoseAtTimstsamp is used to find this one

Reply:
don't know if this is a naming problem or logic problem, but getPoseAtTimestamp should return estimatedPoseAtObservationTime